Add temperature SUPG stabilization

lindsayad opened the pull request:
Stabilization for INS temperature/energy equation, including transient, conduction, and convection terms.

@pbalest I'm going to add the other terms (I believe there are some volumetric source/sink terms that you're interested in), but I thought that I'd get this up here so you can look at it.

Changes which break backwards compatibility:
- Any existing `INSADTemperature` objects have been renamed to `INSADEnergy`
- `INSADEnergyAdvection` now "computes" its residual based on a material property coming from an `INSAD` material. Moreover, it's velocity, cp, and rho parameters have been removed. Consequently the object now perfectly mirrors `INSADMomentumAdvection`
- `INSAD3Eqn` is the material that computes all necessary material properties for simulation of *unstabilized* incompressible Navier-Stokes which includes an energy equation
- `INSADStabilized3Eqn` is the material that computes all necessary material properties for simulation of *stabilized* incompressible Navier-Stokes which includes an energy equation
- `INSADMaterial` and `INSADTauMaterial` retain their original meaning: materials for use with unstabilized and stabilized mass and momentum Navier-Stokes simulations
- Because `INSADEnergyAdvection` now depends on a material property, any simulation that uses that object must have either an `INSAD3Eqn` or `INSADStabilized3Eqn` (or derived) material

Closes #15500

lindsayad committed:
Add temperature SUPG stabilization (#15500)

pbalest commented:
thanks yes I am interested in the source terms too! I'll take a look!

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/553075/) on 51bb068 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15502/site)

This comment will be updated on new commits.

lindsayad committed:
Rework INSADObjectTracker to use InputParameters

This significantly cuts down on the amount of code required in the
INSADObjectTracker class and also makes it easier to add additional
INSAD objects in the future

pbalest commented:
General comment on the naming:

I like the nomenclature the module was following at the beginning for the Kernels:
INSADmass..
INSADmomentum.. 
for the energy equation would be nice having INSADenergy..

what you think about that?

lindsayad commented:
I agree with you but we already had the INS*Temperature convention from
before. It feels bad to change it. I guess I could try and see whether it
breaks any applications that we test

On Fri, Jun 26, 2020 at 12:07 PM Paolo Balestra <notifications@github.com>
wrote:

> General comment on the naming:
>
> I like the nomenclature the module was following at the beginning for the
> Kernels:
> INSADmass..
> INSADmomentum..
> for the energy equation would be nice having INSADenergy..
>
> what you think about that?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/15502#issuecomment-650347795>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4GZJTLDSDXYGYA7OR3RYTWWNANCNFSM4OIYFWWQ>
> .
>

lindsayad committed:
Add INSADTemperatureWallConvection (#15500)

lindsayad committed:
Add INSTemperatureSource object (#15500)

lindsayad commented:
Ok I haven't done the renaming yet but I've added the wall convection and generic heat source objects

pbalest commented:
what about the tau for energy, are you planning to use something like:
![image](https://user-images.githubusercontent.com/35785210/85898131-ef026b00-b7b8-11ea-8017-37b0d299d9a5.png)
Following the discussion we had last time the units within tau are not consistent since instead of nu there is k.

lindsayad committed:
Rename INSADTemperature to INSADEnergy (#15500)

lindsayad commented:
Yes I will add that

lindsayad commented:
This segmentation fault is brutal. I can reproduce it on Rod. It comes out of dereferencing a parent in a red black three that is NULL. The full backtrace:
```
#0  0x00007fffe7d9d06a in std::local_Rb_tree_decrement (__x=0xc96888) at ../../../../../gcc-5.1.0/libstdc++-v3/src/c++98/tree.cc:98
#1  0x00000000004187f1 in std::_Rb_tree_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*> >::operator-- (this=0x7fffffff9700) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/stl_tree.h:220
#2  0x0000000000418753 in std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*> > >::_M_get_insert_unique_pos (this=0xc96880, __k=...) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/stl_tree.h:1829
#3  0x0000000000417885 in std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*> > >::_M_get_insert_hint_unique_pos (this=0xc96880, __position=..., __k=...)
    at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/stl_tree.h:1926
#4  0x0000000000416922 in std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*> > >::_M_emplace_hint_unique<std::piecewise_construct_t const&, std::tuple<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>, std::tuple<> >(std::_Rb_tree_const_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*> >, std::piecewise_construct_t const&, std::tuple<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>&&, std::tuple<>&&) (this=0xc96880, __pos=...)
    at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/stl_tree.h:2184
#5  0x00000000004158f0 in std::__cxx1998::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, libMesh::Parameters::Value*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*> > >::operator[] (this=0xc96880, __k=...) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/stl_map.h:480
#6  0x00007ffff76c152a in InputParameters::set<bool> (this=0xc96860, name=..., quiet_mode=false) at /home/lindad/projects/moose/framework/build/header_symlinks/InputParameters.h:1007
#7  0x00007ffff76b2864 in InputParameters::addParam<bool, bool> (this=0xc96860, name=..., value=@0x7fffffff9d1f: true, doc_string=...) at /home/lindad/projects/moose/framework/build/header_symlinks/InputParameters.h:1217
#8  0x00007ffff7733055 in INSADObjectTracker::INSADObjectTracker (this=0xc96020, parameters=...) at /home/lindad/projects/moose/modules/navier_stokes/src/userobjects/INSADObjectTracker.C:26
#9  0x00007ffff773aaec in __gnu_cxx::new_allocator<INSADObjectTracker>::construct<INSADObjectTracker, InputParameters const&> (this=0x7fffffffa207, __p=0xc96020) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/ext/new_allocator.h:120
#10 0x00007ffff773aa02 in std::allocator_traits<std::allocator<INSADObjectTracker> >::_S_construct<INSADObjectTracker, InputParameters const&> (__a=..., __p=0xc96020) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/alloc_traits.h:256
#11 0x00007ffff773a696 in std::allocator_traits<std::allocator<INSADObjectTracker> >::construct<INSADObjectTracker, InputParameters const&> (__a=..., __p=0xc96020) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/alloc_traits.h:402
#12 0x00007ffff773a2af in std::_Sp_counted_ptr_inplace<INSADObjectTracker, std::allocator<INSADObjectTracker>, (__gnu_cxx::_Lock_policy)2>::_Sp_counted_ptr_inplace<InputParameters const&> (this=0xc96010, __a=...) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/shared_ptr_base.h:522
#13 0x00007ffff7739b20 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<INSADObjectTracker, std::allocator<INSADObjectTracker>, InputParameters const&> (this=0x7fffffffa3a8, __a=...) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/shared_ptr_base.h:617
#14 0x00007ffff77392d2 in std::__shared_ptr<INSADObjectTracker, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<INSADObjectTracker>, InputParameters const&> (this=0x7fffffffa3a0, __tag=..., __a=...) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/shared_ptr_base.h:1097
#15 0x00007ffff7738bb0 in std::shared_ptr<INSADObjectTracker>::shared_ptr<std::allocator<INSADObjectTracker>, InputParameters const&> (this=0x7fffffffa3a0, __tag=..., __a=...) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/shared_ptr.h:319
#16 0x00007ffff773813b in std::allocate_shared<INSADObjectTracker, std::allocator<INSADObjectTracker>, InputParameters const&> (__a=...) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/shared_ptr.h:614
#17 0x00007ffff7737653 in std::make_shared<INSADObjectTracker, InputParameters const&> () at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/shared_ptr.h:630
#18 0x00007ffff7736932 in moose::internal::buildObj<INSADObjectTracker> (parameters=...) at /home/lindad/projects/moose/framework/build/header_symlinks/Registry.h:146
#19 0x00007ffff4ddf455 in Factory::create (this=0x6f0d08, obj_name=..., name=..., parameters=..., tid=0, print_deprecated=false) at /home/lindad/projects/moose/framework/src/base/Factory.C:131
#20 0x00007ffff45369ca in Factory::create<UserObject> (this=0x6f0d08, obj_name=..., name=..., parameters=..., tid=0) at /home/lindad/projects/moose/framework/build/header_symlinks/Factory.h:306
#21 0x00007ffff44f4e9d in FEProblemBase::addUserObject (this=0xab9d80, user_object_name=..., name=..., parameters=...) at /home/lindad/projects/moose/framework/src/problems/FEProblemBase.C:3379
#22 0x00007ffff77cfa7e in INSADMaterial::INSADMaterial (this=0xc31790, parameters=...) at /home/lindad/projects/moose/modules/navier_stokes/src/materials/INSADMaterial.C:60
#23 0x00007ffff77cdc55 in INSAD3Eqn::INSAD3Eqn (this=0xc31790, parameters=...) at /home/lindad/projects/moose/modules/navier_stokes/src/materials/INSAD3Eqn.C:45
#24 0x00007ffff77e1f30 in __gnu_cxx::new_allocator<INSAD3Eqn>::construct<INSAD3Eqn, InputParameters const&> (this=0x7fffffffb767, __p=0xc31790) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/ext/new_allocator.h:120
#25 0x00007ffff77e1ad2 in std::allocator_traits<std::allocator<INSAD3Eqn> >::_S_construct<INSAD3Eqn, InputParameters const&> (__a=..., __p=0xc31790) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/alloc_traits.h:256
#26 0x00007ffff77e137c in std::allocator_traits<std::allocator<INSAD3Eqn> >::construct<INSAD3Eqn, InputParameters const&> (__a=..., __p=0xc31790) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/alloc_traits.h:402
#27 0x00007ffff77e088f in std::_Sp_counted_ptr_inplace<INSAD3Eqn, std::allocator<INSAD3Eqn>, (__gnu_cxx::_Lock_policy)2>::_Sp_counted_ptr_inplace<InputParameters const&> (this=0xc31780, __a=...) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/shared_ptr_base.h:522
#28 0x00007ffff77dfef0 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<INSAD3Eqn, std::allocator<INSAD3Eqn>, InputParameters const&> (this=0x7fffffffb908, __a=...) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/shared_ptr_base.h:617
#29 0x00007ffff77df9ec in std::__shared_ptr<INSAD3Eqn, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<INSAD3Eqn>, InputParameters const&> (this=0x7fffffffb900, __tag=..., __a=...) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/shared_ptr_base.h:1097
#30 0x00007ffff77df5fc in std::shared_ptr<INSAD3Eqn>::shared_ptr<std::allocator<INSAD3Eqn>, InputParameters const&> (this=0x7fffffffb900, __tag=..., __a=...) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/shared_ptr.h:319
#31 0x00007ffff77ded23 in std::allocate_shared<INSAD3Eqn, std::allocator<INSAD3Eqn>, InputParameters const&> (__a=...) at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/shared_ptr.h:614
#32 0x00007ffff77dd3c9 in std::make_shared<INSAD3Eqn, InputParameters const&> () at /opt/min_moose/gcc-5.1.0/include/c++/5.1.0/bits/shared_ptr.h:630
#33 0x00007ffff77dbca6 in moose::internal::buildObj<INSAD3Eqn> (parameters=...) at /home/lindad/projects/moose/framework/build/header_symlinks/Registry.h:146
#34 0x00007ffff4ddf455 in Factory::create (this=0x6f0d08, obj_name=..., name=..., parameters=..., tid=0, print_deprecated=false) at /home/lindad/projects/moose/framework/src/base/Factory.C:131
#35 0x00007ffff4535e00 in Factory::create<MaterialBase> (this=0x6f0d08, obj_name=..., name=..., parameters=..., tid=0) at /home/lindad/projects/moose/framework/build/header_symlinks/Factory.h:306
#36 0x00007ffff44f242d in FEProblemBase::addMaterialHelper (this=0xab9d80, warehouses=..., mat_name=..., name=..., parameters=...) at /home/lindad/projects/moose/framework/src/problems/FEProblemBase.C:3023
#37 0x00007ffff44f1f09 in FEProblemBase::addMaterial (this=0xab9d80, mat_name=..., name=..., parameters=...) at /home/lindad/projects/moose/framework/src/problems/FEProblemBase.C:2982
#38 0x00007ffff42ae909 in AddMaterialAction::act (this=0x9853d0) at /home/lindad/projects/moose/framework/src/actions/AddMaterialAction.C:29
#39 0x00007ffff42a3183 in Action::timedAct (this=0x9853d0) at /home/lindad/projects/moose/framework/src/actions/Action.C:93
#40 0x00007ffff42a82d1 in ActionWarehouse::executeActionsWithAction (this=0x6f0998, task=...) at /home/lindad/projects/moose/framework/src/actions/ActionWarehouse.C:378
#41 0x00007ffff42a7db2 in ActionWarehouse::executeAllActions (this=0x6f0998) at /home/lindad/projects/moose/framework/src/actions/ActionWarehouse.C:339
#42 0x00007ffff4eb96bd in MooseApp::runInputFile (this=0x6ef7f0) at /home/lindad/projects/moose/framework/src/base/MooseApp.C:892
#43 0x00007ffff4eba62f in MooseApp::run (this=0x6ef7f0) at /home/lindad/projects/moose/framework/src/base/MooseApp.C:1057
#44 0x0000000000413da0 in main (argc=3, argv=0x7fffffffd368) at /home/lindad/projects/moose/modules/navier_stokes/src/main.C:36
```
How is it that I'm getting such bad behavior with simple old `std::map::operator[]`...

lindsayad commented:
This happens during `libMesh::Parameters::insert<bool>`:
```c++
_values[name] = new Parameter<T>;
```
where `T = bool` here. `_values` has no elements in it at this point. Why should this operation fail? Throwing a hail mary to @roystgnr here who's the best C++ guy I know. Any ideas here?

lindsayad commented:
Alright I tracked it down to this fellow: https://gcc.gnu.org/legacy-ml/libstdc++/2015-05/msg00142.html

@milljm I might have to ask for a bump of min gcc...

roystgnr commented:
> Alright I tracked it down to this fellow: https://gcc.gnu.org/legacy-ml/libstdc++/2015-05/msg00142.html

Kind of surprised this bug made it into a released compiler if it's general enough to hit us - it applies to *any* 8 byte value type, including our ```Value*```, not just long long?

> Throwing a hail mary to @roystgnr here who's the best C++ guy I know.

I guess it's time to pass on my crown and scepter?  I've hit and worked around compiler bugs in the past, but I've never actually found a *patch* for one.

(Actually, I never got anywhere near the crown - right around the time I was getting decent with C++ was when I met Wolfgang Bangerth, who's on the gcc contributors page)

lindsayad commented:
Hahaha it definitely took a while to track this down. See the updated version of my perusals ^

lindsayad commented:
Hmm this kind of sucks. The first tag that contains the removal of the bad casts is 6.1. I have to say I'm a little baffled that we don't trigger this anywhere in current libMesh/MOOSE. It makes me wonder whether I am doing something wrong here (I have no idea what it would be); otherwise I have no idea what I'm doing differently that we're not doing anywhere else in current libmesh/MOOSE

lindsayad commented:
One thing I'm seeing is that the min gcc version elides the InputParameters copy constructor while our current version does not. I'm wondering if that might be the root of the problem here...

lindsayad commented:
So here's the constructor:
```c++
INSADObjectTracker::INSADObjectTracker(const InputParameters & parameters)                                                                                                                    
    : GeneralUserObject(parameters), InputParameters(emptyInputParameters()), test(emptyInputParameters())
{
[some body stuff]
}
```
Immediately after construction, if I do this in the debugger:
```
gdb) p _values._M_t._M_impl._M_header
$1 = {_M_color = std::_S_red, _M_parent = 0x0, _M_left = 0x7fffffff98f8, _M_right = 0x7fffffff98f8}
(gdb) p &_values._M_t._M_impl._M_header
$2 = (std::_Rb_tree_node_base *) 0xc96888
(gdb) p test._values._M_t._M_impl._M_header
$3 = {_M_color = std::_S_red, _M_parent = 0x0, _M_left = 0xc96a40, _M_right = 0xc96a40}
(gdb) p &test._values._M_t._M_impl._M_header
$4 = (std::_Rb_tree_node_base *) 0xc96a40
```
Based on the red-black tree constructor the address of `_M_header` and `_M_header._M_left` should be the same. This is the result for the member object `test` but not for actual object. I'm wondering if there's some overlap with this guy: https://stackoverflow.com/questions/53883664/copy-elision-for-direct-base-class-initialization. Anyway it looks like I'm going to have to change my class to just have an `InputParameters` member and then do some extra work to override the behavior of `get` and `set`

lindsayad committed:
Use member InputParameters object instead of deriving from it

This seems to resolve the issue with min gcc

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/554978/) on b1cc439 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15502/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15502/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 37799438739b5dcd7e6041b8d97adf0e35a48f2b`

lindsayad committed:
Change tau for energy to use the thermal conductivity

Previously we were using the same tau for the SUPG stabilization for
both the momentum and energy equations. We now base the dissipative
portion of tau_energy off of the thermal conductivity instead of the
viscosity. Units work out properly

Refs #15502

lindsayad commented:
@pbalest The most recent commit calculates tau_energy based off the thermal conductivity instead of the viscosity (e.g. we now use different taus for momentum and energy). I think this caps the initial work you wanted me to do. Let me know if there's anything else.

pbalest commented:
Thanks a lot @lindsayad, I'll take a look today  and test it whit the input file I have. Just a quick check.. do we have something to apply the coupled force to the momentum equation + the stabilization?

lindsayad commented:
What coupled force?

On Tue, Jun 30, 2020 at 8:44 AM Paolo Balestra <notifications@github.com>
wrote:

> Thanks a lot @lindsayad <https://github.com/lindsayad>, I'll take a look
> today and test it whit the input file I have. Just a quick check.. do we
> have something to apply the coupled force to the momentum equation + the
> stabilization?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/15502#issuecomment-651877937>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4AHP3WX62AZ2EHUDOLRZIB4FANCNFSM4OIYFWWQ>
> .
>

lindsayad committed:
Add documentation (#15500)

pbalest commented:
Something I can use to set the momentum in a closed loop to simulate the effect of a pump.. similarly to what Boussinesqueforce do but instead of gravity, you use a coupled vector variable or just a vector function...

lindsayad commented:
> Something I can use to set the momentum in a closed loop to simulate the effect of a pump.. similarly to what Boussinesqueforce do but instead of gravity, you use a coupled vector variable or just a vector function...

Ok I will add that. Do you have a suggestion for what to name it? I'm curious; what variable would you couple in?

lindsayad commented:
I'm just gonna call it `INSADMomentumCoupledForce` unless I hear otherwise from you.

pbalest commented:
Sorry Alex I was in a meeting, INSADMomentumCoupledForce it's exactly what I had in mind, is it gonna be soemthing you can couple with and auxvectorvariable?

lindsayad commented:
Yes you will be able to couple in an aux vector variable. Just keep in mind that if you use an aux, there won't be any derivative/Jacobian information even if that aux is a function of the nonlinear system degrees of freedom.

pbalest commented:
The Idea was to use a constant for steady states, for transient a time dependent term, if we supply a time dependent function instead of an Auxvariable will this make the AD work?

lindsayad commented:
If the aux variable is a constant with respect to nonlinear degrees of freedom, everything fine. But yes I'm going to let the user either supply a coupled vector variable or a vector  function.

moosebuild commented:
Job [Boost enabled, Pthreads only](https://civet.inl.gov/job/555421/) on 42df79b : invalidated by @lindsayad

pbalest commented:
I see Thank this will help during transient calculations!

lindsayad committed:
Add INSADMomentumCoupledForce (#15500)

lindsayad commented:
Alright, `INSADMomentumCoupledForce` added!

pbalest commented:
First results I got looks very promising all the jigsaw behaviour is gone and convergence is way better. 

![image](https://user-images.githubusercontent.com/35785210/86206299-47bf6400-bb29-11ea-8a01-a8b0588744bd.png)

Coupling with NK and get back to you thanks a lot for your work!

lindsayad commented:
Awesome initial news! You’re welcome!

> On Jun 30, 2020, at 10:29 PM, Paolo Balestra <notifications@github.com> wrote:
> 
> ﻿
> First results I got looks very promising all the jigsaw behaviour is gone and convergence is way better.
> 
> 
> 
> Coupling with NK and get back to you thanks a lot for your work!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

lindsayad commented:
@pbalest I need your opinion on sign conventions. Right now in `INSADMomentumCoupledForce` a positive value for a function or coupled variable component represents a momentum sink in that component direction. In `INSADEnergySource` a positive value for the function corresponds to an energy *source*. Should I rename `INSADEnergySource` to `INSADEnergyCoupledForce` and make the sign conventions the same?

pbalest commented:
Ok let me check if I understood correctly:

if I assign a positive value to INSADMomentumCoupledForce in the x direction, if the flow move toward x+ I get some kind of drag (the kernel will remove momentum) or I will accelerate if is moving x- right?

If this is right I think that INSADEnergySource it's perfect how it is right now, what I would change is the behavior of the INSADMomentumCoupledForce to accelerate the flow in the positive direction if it is positive and slow down it if it is negative.

What you think about that?

lindsayad commented:
Your summary is exactly right. Ok I will change the INSADMomentumCoupledForce behavior then

lindsayad committed:
Allow use of coupled variable in INSADEnergySource

Also change sign convention in INSADMomentumCoupledForce to be
consistent with that in INSADEnergySource

Refs #15500

moosebuild commented:
Job [Boost enabled, Pthreads only](https://civet.inl.gov/job/556390/) on 99f657d : invalidated by @lindsayad

fdkong commented on modules/navier_stokes/src/kernels/INSADEnergySUPG.C:
```diff
@@ -0,0 +1,33 @@
+
+ADReal
+INSADEnergySUPG::precomputeQpStrongResidual()
+{
```
Why this is not `precomputeQpResidual`? You use `precomputeQpStrongResidual` in other classes.

lindsayad replied:
This inherits from a different class called `ADKernelSUPG` which itself inherits from `ADKernelStabilized` which changes the function you have to override to `precomputeQpStrongResidual`

fdkong reviewed (dismissed):
Great! Couple everything

jiangwen84 reviewed (commented):
The ``INSAD3Eqn`` derives from ``INSADMaterial``, so it means the energy related properties cannot be defined without mass and momentum.  Would it be better if the energy part becomes a standalone class?

lindsayad commented:
> The `INSAD3Eqn` derives from `INSADMaterial`, so it means the energy related properties cannot be defined without mass and momentum. Would it be better if the energy part becomes a standalone class?

Do you think people will use the Navier-Stokes module just to simulate energy transport? Technically they could do that by supplying the velocity as an auxiliary vector variable. I like the cleanness of only having to put one material in the input file for the three moments: mass, momentum, and energy. But if you think there's a significant number of folks who would just like to do energy transport (and not mass and momentum), then I could split into two materials.

I guess @jiangwen84 even if just you think you'll do standalone energy transport, then that would be a significant enough motivation to split the classes.

pbalest commented:
I the past I have never used energy transport without momentum and and mass equation. Do not think it's a useful feature on the TH side at least.

jiangwen84 commented:
@lindsayad @pbalest thanks for your feedback. For my application, I always solve three equations together. The only reason I bring this up is to see if anyone would like to reuse the code in some cases where the velocities are AuxVariables. Looks like there is no immediate need for that and the way that  it is implemented now is good :-)

lindsayad commented:
Ok sounds good! If we ever get a user-request for it down the road, then we
can make that happen

On Mon, Jul 6, 2020 at 10:40 AM Wen Jiang <notifications@github.com> wrote:

> @lindsayad <https://github.com/lindsayad> @pbalest
> <https://github.com/pbalest> thanks for your feedback. For my
> application, I always solve three equations together. The only reason I
> bring this up is to see if anyone would like to reuse the code in some
> cases where the velocities are AuxVariables. Looks like there is no
> immediate need for that and the way that it is implemented now is good :-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/15502#issuecomment-654374392>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4GUFHCZZMDTITL4BJTR2IEBDANCNFSM4OIYFWWQ>
> .
>

lindsayad commented:
If I don't hear anything from @YaqiWang by the end of today, I'll merge

pbalest commented:
I am trying to test the coupling with Griffin, but I am having issue because [!49](https://gitlab.software.inl.gov/idaholab/griffin/-/merge_requests/49) is not in yet.. I am working offline for now reloading the solution...

jiangwen84 commented on modules/navier_stokes/doc/content/source/kernels/INSADMomentumCoupledForce.md:
```diff
@@ -0,0 +1,13 @@
+# INSADMomentumCoupledForce
```
It seems there is little overlap of this one and body force term? I wonder if there is any particular case to use this term.

lindsayad replied:
What body force term?

jiangwen84 replied:
For example, gravity? if the function is 0 0 -0.98 then it is the same as gravity term, correct?

lindsayad replied:
Hmm you are indeed right that this object has a superset of the capability provided by `INSADGravityForce`. I'm trying to think how I would generalize the design. A user could add multiple `INSADMomentumCoupledForce` objects in their input file. Each would have to use a unique coupled variable or unique function, and then I'd have to create an arbitrary number of coupled force properties in the `INSADMaterial`. Conceptually that seems straightforward enough. More difficult and painstaking would be adding this capability to the new `INSAction`. It would have a parameter like `params.addParam<std::vector<VariableName>>("additional_coupled_variable_forces")` and `params.addParam<std::vector<FunctionName>>("additional_coupled_function_forces")` and then parse those parameters and add the objects.

This would probably be 1-2 days work. Not sure if I have the time for that right now unless Griffin cares about it.

jiangwen84 replied:
I see it becomes a little messy for Action. I would not worry about this for now and we can improve it in the future.

lindsayad replied:
Actually I could still do this with a single object. I would just change the coupled variables/functions parameters to take multiple arguments (the coupled variables parameter already does this anyways)

lindsayad replied:
Hmm @jiangwen84 I think there's probably still value in having the `INSADGravityForce` object. For example in my first naive attempt to simply sub the `INSADMomentumCoupledForce` in the boussinesq test, I applied this patch:
```
diff --git a/modules/navier_stokes/test/tests/ins/boussinesq/boussinesq_square.i b/modules/navier_stokes/test/tests/ins/boussinesq/boussinesq_square.i
index 24c90922f8..8554a00be1 100644
--- a/modules/navier_stokes/test/tests/ins/boussinesq/boussinesq_square.i
+++ b/modules/navier_stokes/test/tests/ins/boussinesq/boussinesq_square.i
@@ -126,9 +126,9 @@
     gravity = '0 -9.81 0'
   [../]
   [./gravity]
-    type = INSADGravityForce
+    type = INSADMomentumCoupledForce
     variable = velocity
-    gravity = '0 -9.81 0'
+    vector_function = 'vector_func'
   [../]
 []
 
@@ -151,3 +151,11 @@
     temperature = temp
   []
 []
+
+[Functions]
+  [vector_func]
+    type = ParsedVectorFunction
+    value_x = '0'
+    value_y = '-9.81'
+  []
+[]
```
Result: exodiff. The reason being that the gravity vector is multiplied by rho.

jiangwen84 replied:
Sounds good to me. Thanks!

jiangwen84 reviewed (dismissed):
Two more questions. Thanks!

lindsayad commented on modules/navier_stokes/doc/content/source/kernels/INSADEnergyWallConvection.md:
```diff
@@ -0,0 +1,16 @@
+# INSADEnergyWallConvection
+
+This object adds a term of the incompressible energy equation of the form
```
I also thought the same thing about boundary vs volume, but @pbalest has his reasons for wanting it this way. Perhaps I'll let  him answer your question

pbalest replied:
This term is necessary if you want to simulate an HX not modeling the full geometry (all the pipes if the flow is within the pipes or the shell if is outside). If you impose it on a surface your geometry will have to respect the actual geometry (to have the same area) without considering that only the fluid that touch that surface will heat up, when instead at the outlet of the HX the mixing make sure that everything is well mixed therefore it can be well approximated with a heat exchange on a volume instead of a surface.

jiangwen84 replied:
Thanks for your explanation and it makes senes. Just a minor thing to say, could we use a better name than the "Wall", which makes me think of boundary immediately. My two cents :-)

lindsayad replied:
Good idea. I'll change the name

lindsayad commented:
> I am trying to test the coupling with Griffin, but I am having issue because [!49](https://gitlab.software.inl.gov/idaholab/griffin/-/merge_requests/49) is not in yet.. I am working offline for now reloading the solution...

Hmm yea that has to be frustrating. Looks like [!59](https://gitlab.software.inl.gov/idaholab/griffin/-/merge_requests/59) was just merged, so now I can update  49 and we can get that in.

lindsayad committed:
Remove untested object since sympy vector MMS is not ready

See sympy/sympy#19212

YaqiWang commented:
I am in middle of something, possibly can find some time tonight for reviewing this. Will that be ok?

lindsayad committed:
WallConvection -> AmbientConvection (#15500)

lindsayad commented:
> I am in middle of something, possibly can find some time tonight for reviewing this. Will that be ok?

Yes that's fine

lindsayad committed:
Adapt INSADMomentumCoupledForce to take multiple vars/functions

This will allow removal of INSADGravityForce

Refs #15500

lindsayad committed:
Test INSADMomentumCoupledForce multiple var/func force capability

Refs #15500

lindsayad committed:
Remove unused rho_name

lindsayad commented:
@YaqiWang do you want to review this? It's fine if you don't.

YaqiWang commented:
I do not want to hold off this. I am busy on something and cannot find a few hours (sounds an execute ;-) It is fine to merge without my review. If issues pop up in the future, we can always fix or improve. Just want to get an answer for one thing: is the action updated with this?

lindsayad commented:
>  is the action updated with this?

Yes sir! I've added a matching action test for every non-action test.

@fdkong can you re-approve please?
